docs/zh-cn/faq_debug.md

```diff
@@ -1,136 +1,136 @@
+如果你在 `rules.mk` 中配置了 `CONSOLE_ENABLE = yes`，你的键盘将会输出调试信息。默认情况下输出很有限，可以启用调试模式来增加调试输出的丰富度。使用你的键映射方案中的 `DEBUG` 键码，或使用[指令](zh-cn/feature_command.md)功能来启动调试模式，或者将下面这段代码放到你的键映射中：
 
-```
-Waiting for new device:.........................
-Listening:
+```c
+void keyboard_post_init_user(void) {
+  // 通过调整这些值可以改变其表现
+  debug_enable=true;
+  debug_matrix=true;
+  //debug_keyboard=true;
+  //debug_mouse=true;
+}
 ```
 
-如果您无法获得这条“Listening:”消息，请尝试在[Makefile]中使用 `CONSOLE_ENABLE=yes`
+## 调试工具
 
-在Linux这样的操作系统上，你可能需要一些权限。
-- 使用`sudo hid_listen`
+有多种可用于调试的工具。
 
-## 控制台没有返回消息
-检查:
-- *hid_listen* 找到了你的设备。看前面。
-- 输入**Magic**+d打开调试。详见[Magic Commands](https://github.com/tmk/tmk_keyboard#magic-commands)。
-- 设置`debug_enable=true` ，一般存在于**matrix.c**的`matrix_init()`中。
-- 尝试使用'print'函数而不要用调试输出。详见**common/print.h**。
-- 断开其他有控制台功能的设备。 详见[Issue #97](https://github.com/tmk/tmk_keyboard/issues/97)。
+### 使用QMK工具箱调试
 
-## Linux或UNIX这样的系统如何请求超级用户权限
-用'sudo'来执行*hid_listen*就有权限了。
-```
-$ sudo hid_listen
-```
+在兼容的平台上，[QMK工具箱](https://github.com/qmk/qmk_toolbox)可以展示你的键盘的调试输出。
 
-或者把一个文件放到规则文件夹来为TMK设备添加*udev规则*，不同系统的目录可能有所不同。
+### 使用 QMK CLI 进行调试
 
-文件: /etc/udev/rules.d/52-tmk-keyboard.rules(在Ubuntu系统的情况下)
-```
-# tmk keyboard products     https://github.com/tmk/tmk_keyboard
-SUBSYSTEMS=="usb", ATTRS{idVendor}=="feed", MODE:="0666"
-```
+倾向于在终端进行调试？使用 [QMK CLI 命令行](zh-cn/cli_commands.md#qmk-console)可以展示键盘输出的调试信息。
 
-***
+### 使用hid_listen调试
 
-# 其他
-## 安全注意事项
+更喜欢使用终端的方案？PJRC提供的[hid_listen](https://www.pjrc.com/teensy/hid_listen.html)也可以用来展示调试信息，已有Windows、Linux及MacOS下预编译好的可执行文件。
 
-你应该不想要把你的键盘变成"砖头"吧，就是变成没法重写固件的那种。
-下面讲解一些参数来告诉你什么风险很大（其实也不是很大）。
+## 发送自定义调试信息 :id=debug-api
 
-- 假如你键盘表面没有设计重置键"RESET", 那你要进入bootloader的话就要按PCB上的RESET了。
-  按PCB上的RESET要拧开键盘底部。
-- 如果 tmk_core / common 里面的文件丢失键盘可能失灵。
-- .hex太大可能不太好; `make dfu` 会删除块，检验大小（咦?好像反了...）。
-  一但出错，刷新键盘失败的话就困在DFU出不去了。
-  - 所以, 要知道大小限制。 Planck键盘上.hex文件最大大小是 is 7000h (十进制是28672)
+有时在[自定义代码](zh-cn/custom_quantum_functions.md)中输出调试信息非常有用，要做到这个功能也很简单，在代码文件头部包含 `print.h` 文件：
 
+```c
+#include "print.h"
 ```
-Linking: .build/planck_rev4_cbbrowne.elf                                                            [OK]
-Creating load file for Flash: .build/planck_rev4_cbbrowne.hex                                       [OK]
 
-Size after:
-   text    data     bss     dec     hex filename
-      0   22396       0   22396    577c planck_rev4_cbbrowne.hex
-```
+然后可以使用以下输出函数：
 
-  - 上面那个文件大小是 22396/577ch，比28672/7000h小
-  - 当你有一个合适的.hex文件时，你就要重试加载那个了
-  - 您在键盘Makefile中的某些选项可能消耗额外内存；注意以下这几个
-    BOOTMAGIC_ENABLE, MOUSEKEY_ENABLE, EXTRAKEY_ENABLE, CONSOLE_ENABLE, API_SYSEX_ENABLE
-- DFU 工具/不/可以写入bootloader (unless you throw in extra fruit salad of options), 
-  所以还是有点危险的
-- EEPROM大概有100000次循环寿命。不要总是频繁重写固件；EEPROM会玩坏的。
-## 全键无冲不好用
-首先你要在**Makefile**用如下命令编译固件`NKRO_ENABLE`。
+* `print("string")`: 字符串输出
+* `uprintf("%s string", var)`: 格式化字符串输出
+* `dprint("string")` 仅调试模式下，字符串输出Print a simple string, but only when debug mode is enabled
```

Comment:
```suggestion
* `dprint("string")` 仅调试模式下，字符串输出
```

Reply:
suggestions flooded in files, I made another commit for these.